File: scripts/lib/fake-binary.rs
//! A stand-in for a shipped Deslop binary. [DEPLOY-VSIX-PACKAGE]
//!
//! The verifier proofs need executables that report a *chosen* name and
//! version, so that a verifier's rejection has something real to reject. A
//! `#!/bin/sh` script cannot be one of those on Windows: there is no shebang,
//! and `CreateProcess` will not start a text file — every proof that staged
//! one got a null exit status and empty output, and then compared that empty
//! output against an expectation and passed.
//!
//! So the fixture is a real program. It is compiled once per test run and
//! copied for each fixture, with that fixture's two answers appended after the
//! marker below — which is why it reads its own image rather than taking the
//! answers from a file beside it: a second file staged next to a binary would
//! show up inside the package fixtures as an undeclared entry, and the
//! verifiers are right to reject those.
//!
//! Built by `scripts/lib/fake-binary.mjs`; it is not part of the workspace.

use std::error::Error;
use std::fs;

/// Separates the compiled image from the answers appended to a copy of it.
const MARKER: &[u8] = b"\n@@DESLOP-FAKE-BINARY-PAYLOAD@@\n";

/// Prints the JSON answer for `--version --json` and the plain one otherwise,
/// matching the two shapes `[DEPLOY-BINARY-VERSION]` requires of a real binary.
fn main() -> Result<(), Box<dyn Error>> {
    let image = fs::read(std::env::current_exe()?)?;
    let start = last_offset(&image, MARKER).ok_or("this copy carries no appended answers")?;
    let answers = String::from_utf8(image[start + MARKER.len()..].to_vec())?;
    let mut lines = answers.lines();
    let plain = lines.next().ok_or("the appended answers have no plain version line")?;
    let json = lines.next().ok_or("the appended answers have no JSON version line")?;
    let wants_json = std::env::args().any(|argument| argument == "--json");
    println!("{}", if wants_json { json } else { plain });
    Ok(())
}

/// Offset of the LAST occurrence of `needle`. The marker is also a string
/// constant inside the compiled image, so the first occurrence is this
/// program's own code and the appended copy is always the final one.
fn last_offset(haystack: &[u8], needle: &[u8]) -> Option<usize> {
    haystack.windows(needle.len()).rposition(|window| window == needle)
}

File: scripts/lib/fake-binary.mjs
// Fixture binaries that really are binaries.
// [DEPLOY-GATE-PORTABILITY] [DEPLOY-VSIX-PACKAGE]
//
// Every deployment verifier ends by running the artifact it is checking and
// reading what it prints, so a proof that a verifier rejects a drifted binary
// has to stage something a verifier can actually run. A shell script is that
// on Unix and nothing at all on Windows, where `CreateProcess` starts an image
// and not a text file — so on Windows each proof spawned nothing, read empty
// output, and passed anyway.
//
// The fixture is therefore compiled, once, from `fake-binary.rs`. `rustc` is
// the one compiler this repository is guaranteed to have: the deployment gate
// runs after the workspace build, and the CI job that runs it installs the
// Rust toolchain for exactly that reason. Each fixture is a copy of that one
// build with its own two answers appended, so staging thirty of them costs one
// compile.

import { spawnSync } from "node:child_process";
import { appendFileSync, chmodSync, copyFileSync, mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

import { currentPlatform, executableName } from "../release/vsix-platforms.mjs";

/** Must equal `MARKER` in fake-binary.rs, byte for byte. */
const PAYLOAD_MARKER = "\n@@DESLOP-FAKE-BINARY-PAYLOAD@@\n";

/** Mode a staged fixture carries, matching what a release build produces. */
const EXECUTABLE_MODE = 0o755;

/** Optimised and stripped: every fixture is copied into a zip several times. */
const RUSTC_ARGUMENTS = ["--edition", "2021", "-O", "-C", "strip=symbols"];

const SOURCE = fileURLToPath(new URL("fake-binary.rs", import.meta.url));

let compiled;

/**
 * Path to the compiled fixture program, building it on first use.
 *
 * @returns {string} absolute path to an executable for this host
 */
export function fakeBinaryTemplate() {
  if (compiled !== undefined) return compiled;
  const directory = mkdtempSync(join(tmpdir(), "deslop-fake-binary-"));
  const output = join(directory, executableName("fake", currentPlatform()));
  const result = spawnSync("rustc", [...RUSTC_ARGUMENTS, "-o", output, SOURCE], {
    encoding: "utf8",
  });
  if (result.status !== 0) {
    throw new Error(`rustc could not build ${SOURCE}: ${result.stderr ?? result.error}`);
  }
  compiled = output;
  return compiled;
}

/**
 * Stages a runnable fixture at `path` that answers `--version` with
 * `answers.plain` and `--version --json` with `answers.json`.
 *
 * @param {string} path where to write the fixture, suffix included
 * @param {{plain: string, json: object}} answers what the fixture reports
 * @returns {string} `path`
 */
export function writeFakeBinary(path, answers) {
  copyFileSync(fakeBinaryTemplate(), path);
  const lines = [answers.plain, JSON.stringify(answers.json)];
  appendFileSync(path, `${PAYLOAD_MARKER}${lines.join("\n")}\n`);
  chmodSync(path, EXECUTABLE_MODE);
  return path;
}
